lerp_derive/src/derive.rs

```diff
@@ -0,0 +1,61 @@
+use proc_macro2::TokenStream;
+use quote::{quote};
+use syn::{Error, Field, Fields, Index, ItemStruct, Result, spanned::Spanned};
+
+pub fn lerp_derive_internal(input: ItemStruct) -> Result<TokenStream> {
+    let name = &input.ident;
+
+    match input.fields {
+        Fields::Named(fields) => {
+            if let Some(Field { ty, .. }) = fields.named.first() {
+                let fields = fields.named.iter().map(|f| {
+                    if let Some(name) = f.ident.as_ref() {
+                        Ok(quote! {
+                            #name: self.#name.lerp(other.#name, t)
+                        })
+                    } else {
+                        Err(Error::new(f.ident.span(), "All fields must be named in the struct"))
+                    }
+                }).collect::<Result<Vec<_>>>()?;
+
+                Ok(quote! {
+                    #[automatically_derived]
+                    impl Lerp<#ty> for #name {
+                        fn lerp(self, other: Self, t: #ty) -> Self {
```

Comment:
Extracting `ty` from the first field of the input struct is clever, but restrictive: it sharply limits the structs on which this derive macro can be applied. I think we can be more general without sacrificing accuracy. Something like this (pseudocode, I haven't worked out all the details):

```rust
impl Lerp<f64> for #name {
  fn lerp(self, other: Self, t: f64) -> Self {
    Self {
      // for each #field in #fields, precompute this:
      match #field.ty {
        f64 => #name: self.#name.lerp(other.#name, t),
        f32 => #name: self.#name.lerp(other.#name, t as f32),
        T if T: Lerp<f64> => #name: self.#name.lerp(other.#name, t),
        _ => // compile error
      }
    }
  }
}

// separate impl Lerp<f32> follows, using the same pattern
```

This pattern would allow mixing arbitrary numeric primitives in a struct for which we `derive(Lerp)`, which makes the macro much more powerful. I would prefer that.

Adding duplicate impls for `f64` and `f32` is just a convenience thing, but I can't think of any reason we'd want not to do it.

Reply:
That makes sense, Ill see what I can do